### PR TITLE
Cookie branch not ready to push yet

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**

### CODE REVIEW

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java [hunk 1]
```diff
@@ -50,12 +51,11 @@
...
     @CrossOrigin(origins = "http://localhost:3000")
     @PostMapping("/signin")
-    public String authenticateUser(LoginRequest loginRequest, HttpServletRequest request,
-    HttpServletResponse response) {
-      response.addHeader("Access-Control-Allow-Origin", "http://localhost:3000");
-      response.setHeader("Access-Control-Allow-Credentials", "true");
-      response.setHeader("Access-Control-Allow-Methods", "POST, GET");
-      response.setHeader("Access-Control-Allow-Headers", "Authorization, Content-Type");
+    public ResponseEntity<?> authenticateUser(@Valid @RequestBody LoginRequest loginRequest, HttpServletResponse response) {
+        response.setHeader("Access-Control-Allow-Origin", "http://localhost:3000");
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: @gngster As spoken, here should be setHeader instead of addHeader. It seems response comes with a * header, so when you call addHeader, you have both * and http://localhost:3000 in the headers, which cause an issue.

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java [hunk 2]
```diff
@@ -50,12 +51,11 @@
 
     @CrossOrigin(origins = "http://localhost:3000")
     @PostMapping("/signin")
-    public String authenticateUser(LoginRequest loginRequest, HttpServletRequest request,
-    HttpServletResponse response) {
-      response.addHeader("Access-Control-Allow-Origin", "http://localhost:3000");
-      response.setHeader("Access-Control-Allow-Credentials", "true");
-      response.setHeader("Access-Control-Allow-Methods", "POST, GET");
-      response.setHeader("Access-Control-Allow-Headers", "Authorization, Content-Type");
+    public ResponseEntity<?> authenticateUser(@Valid @RequestBody LoginRequest loginRequest, HttpServletResponse response) {
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: @gngster previously you removed this @Valid @RequestBody, which causes an error also.

Also, cause the front-end is configured to receive a ResponseEntity<?>, changing it to String causes an issue.

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java [hunk 3]
```diff
@@ -68,14 +68,10 @@ public String authenticateUser(LoginRequest loginRequest, HttpServletRequest req
...
-          newCookie.setMaxAge(24 * 60 * 60);
-          response.addCookie(newCookie);
-          return "blabla";
-      /*
-        SecurityContextHolder.getContext().setAuthentication(authentication);
-        String jwt = tokenProvider.generateToken(authentication);
+        newCookie.setMaxAge(24 * 60 * 60);
+        response.addCookie(newCookie);
+
         return ResponseEntity.ok(new JwtAuthenticationResponse(jwt));
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: @gngster @gjlam95 both of you need to decide what to send back to the front-end now. I'm thinking of sending a simple new ResponseEntity<>(HttpStatus.ok), but front-end needs to configure that.

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/config/SecurityConfig.java [hunk 4]
```diff
@@ -66,8 +66,8 @@ public PasswordEncoder passwordEncoder() {
     @Override
     protected void configure(HttpSecurity http) throws Exception {
         http
-                .cors()
-                    .and()
+              //  .cors()
+              //      .and()
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/config/SecurityConfig.java]: Should just delete instead of comment out, since we are using a VCS alrd.

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/config/SecurityConfig.java [hunk 5]
```diff
@@ -100,4 +100,4 @@ protected void configure(HttpSecurity http) throws Exception {
         http.addFilterBefore(jwtAuthenticationFilter(), UsernamePasswordAuthenticationFilter.class);
 
     }
-}
\ No newline at end of file
+}
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/config/SecurityConfig.java]: Unnecessary change

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java (hunk 3 above)
Comment (gjlam95) [reply backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: what is a "simple new ResponseEntity<>(HttpStatus.ok)"?

Comment (Zhiyuan-Amos) [reply backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: Currently, the response body contains the value of the JWT token, which I think is being parsed by front-end code. This `HttpStatus.ok` has no body, so I think there's some conflict with front-end code because it's expecting something to parse.

`HttpStatus.ok` basically means... it's ok, I think response code 200?.

File: backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java [hunk 6]
```diff
@@ -46,19 +49,28 @@
     @Autowired
     JwtTokenProvider tokenProvider;
 
+    @CrossOrigin(origins = "http://localhost:3000")
```
Comment (Zhiyuan-Amos) [inline backend-server/src/main/java/org/cs4239/team1/protectPMLeefrontendserver/controller/AuthController.java]: `http://localhost:3000` should be on `local-development` branch; `master` branch should contain the actual url.

**Comment by gngster** [comment]
@gjlam95 hi can help review? this is ready to push but i will only push it after the bluetooth tag is pushed

File: healthcare-app-client/src/user/login/Login.js [hunk 7]
```diff
@@ -55,7 +55,7 @@ class LoginForm extends Component {
         };
         login(loginRequest)
         .then(response => {
-            localStorage.setItem(AUTH_TOKEN, response.accessToken);
+            localStorage.setItem(AUTH_TOKEN, response.sessionId);
```
Comment (gjlam95) [inline healthcare-app-client/src/user/login/Login.js]: I only looked at Login.js & APIUtils.js since those are the front-end components, just want to clarify that even if an attacker manages to steal `response.sessionId`, it's encrypted is it?

Comment (gngster) [reply healthcare-app-client/src/user/login/Login.js]: it's not encrypted. but even if he steals it somehow he cant use that in any way. the sessionId is stored in the JWT which is in the cookie. so the server will check that this sessionId which is in the request header matches with the sessionID in the jwt in the cookie to ensure no CSRF. if let's say there is XSS and someone gets this sessionID, if he manipulates it, the server will see that it has been manipulated alr when it checks, and the server will say unauthenticated

**Comment by gngster** [comment]
sorry i realised a lot of things have to be changed first... @Zhiyuan-Amos 
we should be encrypting the data in the cookie, sorryyy later need review again @gjlam95

**Comment by gngster** [comment]
Hi @Zhiyuan-Amos , I implemented the encrypt and decrypt of JWT in the cookie, could you take a look at my commits and clean up if you want to :) I need you to help with 2 things - 
1. check out the encryption key that is in plaintext in AuthController & JWTTokenProvider  - could you help me store that in an encrypted way?
2. The encryption requires a IV (that should be re-generated randomly to use in encryption when the JWT is no longer valid). I was thinking that instead of the "sessionId" that you generated randomly to pass to the user, we can instead keep track of the IV inside of the JWT + pass to the user. So the IV will replace the current "sessionId" and then we can also grab this IV from the request header "session-id" for our decryption process. 
So, request comes in from client -> server grabs IV from "session-id" header -> server decrypts the cookie value (aka encrypted JWT) using the encryption key (which is the same all the way) + the IV -> server validates for JWT + checks that IV in JWT == IV in request header -> user signs in
So in the TODO in AuthController, i already included the code there to generate the IV randomly. wondering if you can help me use that code to replace the current "sessionId" code? Let me know also if you agree with this point of replacing "sessionId" with IV. it is safe for IV to be in plaintext.

**Comment by gngster** [comment]
and for both @gjlam95 and @Zhiyuan-Amos I was reading and realised that the cookie should be a session cookie (meaning that it disappears once the user closes the browser - aka means the next time, the user must log in again) instead of a persistent cookie for which we set the expiration date/ maxAge, because that is not as secure. The cookie should not persist after browser is closed. So when user closes browser, user should be forgotten, so they say to do that, just dont put any expiry date or max age on the cookie. But, somehow idk why is it cos we running on localhost or what like the cookie still in the memory but i checked my cookies via chrome it says "cookie expires once browser is closed". But anyway yea, so just to update you all on the security aspects of the cookie - will write everything down in the documentation doc

**Comment by gngster** [comment]
Hi @Zhiyuan-Amos i have looked through the changes you made already, thanks!! I made a few additional changes - you can help me clean up. Because I realised that the cookie value needs to change/ cookie needs to be deleted when the user logs out. if not the user would attempt to log back in with old cookie but fail. But i realised because the cookie is set to HTTPOnly, the javascript side cannot touch it so we have to make a api call to the server when the user wants to log out. so that the server will overwrite the old cookie value with the value "1". and the authenticationfilter will check if the value of the cookie is "1" then we will issue user with new JWT.

**Comment by Zhiyuan-Amos** [comment]
> so that the server will overwrite the old cookie value with the value "1". and the authenticationfilter will check if the value of the cookie is "1" then we will issue user with new JWT.

HAHA what a workaround
